### ISSUE TITLE
Python Driver : Whats the difference between execCypher() and cypher() in age.py

### ISSUE THREAD

**Issue by Munmud** [apache/age#831]
Whats the difference between `execCypher()` and `cypher()` used in age.py in python driver

I found no instruction regarding cypher() but found some comments regardiing execCypher(). See below
```py
# Execute cypher statement and return cursor.
# If cypher statement changes data (create, set, remove), 
# You must commit session(ag.commit()) 
# (Otherwise the execution cannot make any effect.)
```

I am going through the `test_age_py.py` file and in testCollect() I found these 2 function both for retriving the node data
### Using cypher()
```py
        print(" - COLLECT 1 --------")
        with ag.connection.cursor() as cursor:
            ag.cypher(cursor, 
                """
                    MATCH (a)-[:workWith]->(c) 
                    WITH a as V, 
                    COLLECT(c) as CV 
                    RETURN V.name, CV
                """, cols=["V","CV"])
            for row in cursor:
                nm = row[0]
                collected = row[1]
                print(nm, "workWith", [i["name"] for i in collected])
                self.assertEqual(3,len(collected))
```

### Using execCypher()
```py
        print(" - COLLECT 2 --------")
        for row in ag.execCypher(
        """
            MATCH (a)-[:workWith]->(c) 
            WITH a as V, 
            COLLECT(c) as CV 
            RETURN V.name, CV
        """
        , cols=["V1","CV"]):
            nm = row[0]
            collected = row[1]
            print(nm, "workWith", [i["name"] for i in collected])
            self.assertEqual(3,len(collected))
```

**Comment by im-aditya30** [comment]
The `execCypher()` function uses a PostgreSQL prepared statement to execute the Cypher query, while `cypher()` builds a PostgreSQL statement that performs the Cypher query using the `buildCypher()` function. Moreover `execCypher()` is more robust in terms of error handling, while `cypher()` is more customizable and flexible, allowing for more control over the details of the query execution.

**Comment by safi50** [comment]
This `execCypher()` function executes a single Cypher query, passed as a string parameter, and return the result. It manages the connection and cursor creation/closure internally.

On the other hand, `cypher()` creates a reusable Age object that you can use to execute multiple Cypher queries within the same connection. The cypher() function takes a connection object (created using `psycopg2.connect()`) and the graph name as arguments. 

Using `cypher()`, you don't need to manage connections and cursors explicitly.

**Comment by titoausten** [comment]
| Function        | Description                                                                                              |
|-----------------|----------------------------------------------------------------------------------------------------------|
| `Cypher()`      | Takes a single argument, which is the Cypher query string to be executed. Creates a Cypher query object that can be used to build a Cypher query string. It does not execute the query itself.                     |
| `execCypher()`   | Takes a Cypher query object created by the Cypher() function and executes the Cypher query against an AGE graph database. Returns a result object with query results. |

**Comment by moaz178** [comment]
The cypher() method sends a Cypher query to the Neo4j server and returns a cursor object that allows you to iterate over the results of the query. The cursor object is similar to a generator that generates rows of data as you iterate over it. The cypher() method is useful if you want to process the results of a query row-by-row,meaning, if you have a large dataset and want to avoid loading it all into memory at once.

The execCypher() method sends a Cypher query to the Neo4j server and returns the result of the query as a Python object. The type of the Python object depends on the type of the result returned by the Cypher query. For example, if the query returns a single value, execCypher() returns that value as a Python scalar type. If the query returns a set of values, execCypher() returns a list of dictionaries, where each dictionary represents a row of the result set and maps column names to values.

**Comment by M4rcxs** [comment]
### Cypher()

This module takes a **single** argument, which is the Cypher query string to be executed, and returns the results as a list of records. Each record is a dictionary object that maps the names of the columns in the result to their corresponding values.
### execCypher()

This function  executes the Cypher query string using a **cursor**, and then closes the cursor and connection **before** returning the result. Using a cursor to execute the query allows for efficient retrieval of query results and can reduce memory consumption in cases where the result set is large.

**Comment by kenwoon** [comment]
The main difference between execCypher() and cypher() in age.py is in how they execute the Cypher query and handle the result.

The cypher() function creates a Cypher query object that can be used to build a Cypher query string. It takes a connection object and a graph name as arguments and sends the Cypher query to the Neo4j server. It returns a cursor object that allows for iterating over the results of the query row-by-row. This method is useful for processing large datasets and avoiding loading them all into memory at once.

On the other hand, the execCypher() function takes a Cypher query string as input and executes it against an AGE graph database. It creates a cursor internally and returns the result of the query as a Python object. This function is more robust in terms of error handling and allows for efficient retrieval of query results while reducing memory consumption, especially when the result set is large.

**Comment by mahinash26** [comment]
There are two functions for executing Cypher queries in PostgreSQL: execCypher() and cypher().

The execCypher() function uses a PostgreSQL-prepared statement to execute the Cypher query. This method is more robust regarding error handling because it detects and handles errors better than cypher().

The cypher() function, on the other hand, builds a PostgreSQL statement that performs the Cypher query using the buildCypher() function. This approach is more customizable and flexible as it allows for more control over the details of the query execution.

Simply put, execCypher() is a safer and more reliable method for executing Cypher queries, while cypher() gives users more control and flexibility over how the queries are executed.
